### PR TITLE
Unify frame and locals collection implementation between ThreadClient and servers

### PR DESCRIPTION
**Pull request by st0012**
1. Use `get_frame` to retrieve frames with the given frame index.
2. Use `collect_locals` to retrieve the locals from the given frame.
    - This also fixes an issue that `%self` isn't displayed in DAP when stepping backward. 

Thanks to the DAP tests written by @ono-max, I feel a lot comfortable doing such refactoring now 👍

### CODE REVIEW

**Comment by ono-max** [comment]
Could you wait for refactoring until I'm done creating the test framework for CDP?
Thanks.

**Comment by st0012** [comment]
@ono-max I can drop the changes for CDP server if you concern about the test coverage.

**Comment by ono-max** [comment]
That sounds good. Could you drop it? Thank you.

**Comment by st0012** [comment]
@ono-max done 👍

**Comment by ko1** [comment]
why dap tests are changed?

File: test/dap/step_back_test.rb [hunk 1]
```diff
@@ -607,11 +607,19 @@ def test_step_back_works_correctly
             message: "Success",
             body: {
               variables: [
+                {
+                  name: "%self",
```
Comment (st0012) [inline test/dap/step_back_test.rb:611]: @ko1 When stepping forward, `variables` responses contain both `%self` and `bar`. But currently when stepping backward, `%self` is missing for some reason.
This PR happened to make it include `%self` and `bar` when stepping backward as well. I think it's a reasonable behavior so I adopted the tests with it.

**Comment by ono-max** [comment]
Could you rebase these changes so that you can confirm if they work in CI?

**Comment by st0012** [comment]
@ono-max rebased and it passed 👍

**Comment by ono-max** [comment]
Sorry, if you're ok, could you use the test generator instead of modifying tests to rewrite tests?

**Comment by st0012** [comment]
@ko1 I hope this can also be merged first if possible. I'll add more variables tests with #559

**Comment by ono-max** [comment]
Did you fix tests with the test generator?

**Comment by st0012** [comment]
No. I plan to use the new `assert_locals_result` (or whatever you plan to call it) to test it.
The whole idea of adding unit-tests is that I don't need to rerun integration tests for changes like this.

**Comment by ko1** [comment]
Sorry for absent. I'll merge it and I revisit to consider about the name `get_frame`

**Comment by st0012** [comment]
@ko1 thank you 🙏

**Pull request merged**